packages/fxa-content-server/app/scripts/lib/metrics.js

```diff
@@ -344,7 +348,28 @@ _.extend(Metrics.prototype, Backbone.Events, {
     // reported again.
     this._numStoredAccounts = '';
 
-    this._validateSanitizeUtmParams(filteredData);
+    // Create a sanitizer and check the data. This will report issues to sentry
+    // and keep track of critical errors that may have been encountered.
+    const reporter = new MetricErrorReporter(this._sentryMetrics);
+    const validator = new MetricValidator(reporter, {
+      maxEventOffset: this._maxEventOffset,
+      // These are optional, but we will do this to ensure backwards compatibility
+      isUtmValid: Validate.isUtmValid,
+      isDeviceIdValid: Validate.isDeviceIdValid,
+    });
+
+    validator.sanitizeDeviceId(filteredData);
```

Comment:
I'm probably missing something or need to stare at this longer, but do we need to resave the data returned from these method calls before we `this._send(filteredData)` further down? On a similar note, why send in the entire object and pull out pieces in the method rather than send in only the piece of data you need sanitized to each one?

Reply:
We do not need to re-save the data. The object provided is mutated, so the return value is really only a convenience. This decision is more of a by-product of how the code evolved than anything. There was already an instance of the data being mutated in place, so I continued this pattern. I suppose the return value isn't actually needed, although I do use it to make the tests more compact. I guess I’ll remove the return value so it's clear this isn't some sort of pure function.
 
IMO, passing the entire object in this case is okay, because it keeps the inputs to these sanitizing calls consistent. By the way, note the types of the sanitize methods. Each is a sub type of filtered data. This means the function still indicates a minimal set of data required and can be tested or used in isolation without requiring the full filtered data object. In the future it's also possible that we might encounter a scenario where one of these methods sanitizes a large number of fields on the filteredData object. By setting it up this way, we just keep the calls consistent, and then make the subtype required for the sanitizing function more broad. Anyways, I hope that explains some of the reasons I coded it like this.

Reply:
Cool yeah, it looks cleaner to me and more intentional without those returns. 👌

SGTM too on reason for passing in the entire object, I agree it's nice that they all take in the same data, I mostly asked that mostly for my own learning purposes since I tend to pass in minimally needed data and was curious of your reasoning.